src/ring_theory/finiteness.lean

```diff
@@ -550,3 +551,58 @@ ring_hom.finite_presentation.of_finite_type
 end finite_presentation
 
 end alg_hom
+
+section monoid_algebra
+
+namespace add_monoid_algebra
+
+open algebra submodule
+
+variables {R : Type*} {M : Type*} [comm_ring R] [add_comm_monoid M]
+
+lemma ft_of_fg : add_monoid.fg M → finite_type R (add_monoid_algebra R M) :=
+begin
+  rintro ⟨S, hS⟩,
+  rw finite_type.iff_quotient_mv_polynomial',
+  let S₁ := ulift {x // x ∈ S},
+  let ψ := λ (s : S₁), of R M s.down.1,
+  refine ⟨S₁, by apply_instance, mv_polynomial.aeval ψ, λ f, finsupp.induction_linear f _ _ _⟩,
+  { exact ⟨0, by simp only [alg_hom.map_zero]⟩ },
+  { rintro f g ⟨P, hP⟩ ⟨Q, hQ⟩,
+    use P + Q,
+    simp only [ψ, of_apply, subtype.val_eq_coe] at hP hQ,
+    simp only [hP, hQ, ψ, alg_hom.map_add, of_apply, subtype.val_eq_coe] },
+  { intros m r,
+    suffices : ∃ (P : mv_polynomial S₁ R), (mv_polynomial.aeval ψ) P = finsupp.single m 1,
+    { obtain ⟨P, hP⟩ := this,
+      exact ⟨r • P, by simp only [hP, alg_hom.map_smul, mul_one, finsupp.smul_single']⟩ },
+    have : m ∈ span ℕ (S : set M),
+    { rw [← span_nat_eq_add_submonoid_closure, ← @top_to_add_submonoid ℕ M _ _ _] at hS,
+      simp only [to_add_submonoid_injective hS] },
```

Comment:
We're missing a lemma here:
```lean
@[simp] theorem _root_.submodule.mem_to_add_submonoid {R M} [semiring R] [add_comm_monoid M] [module R M] (p : submodule R M) (x : M):
  x ∈ p.to_add_submonoid ↔ x ∈ p  := iff.rfl
```
which lets this become the `_`-free:

```suggestion
    { rw [←submodule.mem_to_add_submonoid, submodule.span_nat_eq_add_submonoid_closure, hS],
      exact add_submonoid.mem_top _ },
```

Reply:
I think @jcommelin  added `submodule.mem_to_add_subgroup` very recently, but I guess didn't need this one.